File: python/cellgen/serialise.py
"""
Serialise a CellSet back to a CellGen format string.
"""

from __future__ import annotations
from .types import Cell, CellSet, CellularElement, ChromosomeNode, EntityNode


_BARE_LABEL_CHARS = set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_.:-")


def to_cellgen(cell_set: CellSet) -> str:
    """Convert a CellSet back to a CellGen format string."""
    return " ; ".join(_cell_str(c) for c in cell_set.cells)


def _cell_str(cell: Cell) -> str:
    return ", ".join(_replicon_str(r) for r in cell.replicons)


def _replicon_str(r: CellularElement) -> str:
    if isinstance(r, ChromosomeNode):
        inner = ", ".join(_entity_str(m) for m in r.children)
        attrs = _attrs_str(r.attributes)
        return f"({inner}){_label_str(r.label)}{attrs}"
    else:
        return _entity_str(r)


def _entity_str(m: EntityNode) -> str:
    inner = ", ".join(_entity_str(c) for c in m.children)
    attrs = _attrs_str(m.attributes)
    return f"{{{inner}}}{_label_str(m.label)}{attrs}"


def _attrs_str(attrs: dict[str, str]) -> str:
    if not attrs:
        return ""
    pairs = ", ".join(f'{k}="{_escape(v)}"' for k, v in attrs.items())
    return f"[{pairs}]"


def _escape(value: str) -> str:
    return (value.replace("\\", "\\\\").replace('"', '\\"')
            .replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t"))


def _label_str(label: str) -> str:
    if not label:
        return ""
    if all(char in _BARE_LABEL_CHARS for char in label):
        return label
    return f'"{_escape(label)}"'

File: webapp/src/cellgen/serialise.ts
/**
 * Serialise a CellSet back to a CellGen format string.
 */

import type { Cell, CellSet, CellularElement, EntityNode } from './types'

export function toCellGen(cellSet: CellSet): string {
  return cellSet.cells.map(cellStr).join(' ; ')
}

function cellStr(cell: Cell): string {
  return cell.replicons.map(repliconStr).join(', ')
}

function repliconStr(r: CellularElement): string {
  if (r.kind === 'chromosome') {
    const inner = r.children.map(entityStr).join(', ')
    const attrs = attrsStr(r.attributes)
    return `(${inner})${r.label}${attrs}`
  }
  return entityStr(r)
}

function entityStr(m: EntityNode): string {
  const inner = m.children.map(entityStr).join(', ')
  const attrs = attrsStr(m.attributes)
  return `{${inner}}${m.label}${attrs}`
}

function attrsStr(attrs: Record<string, string>): string {
  const pairs = Object.entries(attrs)
  if (!pairs.length) return ''
  return '[' + pairs.map(([k, v]) => `${k}="${v}"`).join(', ') + ']'
}
